### PR TITLE
Refactoring CacheInterface and add early expiration

### PR DESCRIPTION
**Pull request by devanych**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #42

### CODE REVIEW

**Comment by samdark** [comment]
That should be mentioned in readme.

File: src/Metadata/CacheItem.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+use function time;
+
+use const PHP_INT_MAX;
+
+/**
+ * CacheItem store the metadata of cache item.
+ *
+ * @internal
+ */
+final class CacheItem
```
[the remark this answers is not in the document]
Comment (devanych) [reply src/Metadata/CacheItem.php:25]: Added

File: src/Metadata/CacheItems.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+namespace Yiisoft\Cache\Metadata;
+
+use Psr\SimpleCache\CacheInterface;
+
+/**
+ * CacheItems store the metadata of each cache item.
+ *
+ * @internal
+ */
+final class CacheItems
```
[the remark this answers is not in the document]
Comment (devanych) [reply src/Metadata/CacheItems.php:14]: Added

File: src/Metadata/CacheItem.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+{
+    private string $key;
+    private ?int $expiry;
+    private ?Dependency $dependency;
+    private float $updated;
+
+    /**
+     * @param string $key The key that identifies the cache item.
+     * @param int|null $expiry The cache expiry or null that meaning infinity.
```
Comment (samdark) [inline src/Metadata/CacheItem.php:34]:
```suggestion
     * @param int|null $expiry The cache expiry. null means infinity.
```

File: src/Metadata/CacheItem.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+     * @param Dependency|null $dependency The cache dependency or null if it is not assigned.
```
Comment (samdark) [inline src/Metadata/CacheItem.php:35]:
```suggestion
     * @param Dependency|null $dependency The cache invalidation dependency or null for none.
```

File: src/Metadata/CacheItem.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        $this->key = $key;
+        $this->expiry = $expiry;
+        $this->dependency = $dependency;
+        $this->updated = microtime(true);
+    }
+
+    /**
+     * Updates the metadata of the cache item.
+     *
+     * @param int|null $expiry The cache expiry or null that meaning infinity.
```
Comment (samdark) [inline src/Metadata/CacheItem.php:48]:
```suggestion
     * @param int|null $expiry The cache expiry. null means infinity.
```

File: src/Metadata/CacheItem.php [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+     * @param Dependency|null $dependency The cache dependency or null if it is not assigned.
```
Comment (samdark) [inline src/Metadata/CacheItem.php:49]:
```suggestion
     * @param Dependency|null $dependency The cache invalidation dependency or null for none.
```

File: src/Metadata/CacheItem.php [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+     *
+     * @return string The key that identifies the cache item.
+     */
+    public function key(): string
+    {
+        return $this->key;
+    }
+
+    /**
+     * Returns a cache expiry or null that meaning infinity.
```
Comment (samdark) [inline src/Metadata/CacheItem.php:69]:
```suggestion
     * Returns a cache expiry or null that means infinity.
```

File: src/Metadata/CacheItem.php [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+     *
+     * @return int|null The cache expiry or null that meaning infinity.
```
Comment (samdark) [inline src/Metadata/CacheItem.php:71]:
```suggestion
     * @return int|null The cache expiry timestamp or null that means infinity.
```

File: src/Metadata/CacheItem.php [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    public function expiry(): ?int
+    {
+        return $this->expiry;
+    }
+
+    /**
+     * Returns a cache dependency or null if it is not assigned.
```
Comment (samdark) [inline src/Metadata/CacheItem.php:79]:
```suggestion
     * Returns a cache dependency or null if there is none.
```

File: src/Metadata/CacheItem.php [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+     *
+     * @return Dependency|null The cache dependency or null if it is not assigned.
```
Comment (samdark) [inline src/Metadata/CacheItem.php:81]:
```suggestion
     * @return Dependency|null The cache dependency or null if there is none.

File: src/Metadata/CacheItem.php [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    public function dependency(): ?Dependency
+    {
+        return $this->dependency;
+    }
+
+    /**
+     * Checks whether the dependency has been changed and whether the cache expires.
```
Comment (samdark) [inline src/Metadata/CacheItem.php:89]:
```suggestion
     * Checks whether the dependency has been changed or whether the cache expired.
```

File: src/Metadata/CacheItem.php [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+     *
+     * @param float $beta The value for calculating the range that is used for "Probably early expiration" algorithm.
+     * @param CacheInterface $cache The actual cache handler.
+     *
+     * @return bool Whether the dependency has been changed and whether the cache expires.
```
Comment (samdark) [inline src/Metadata/CacheItem.php:94]:
```suggestion
     * @return bool Whether the dependency has been changed or whether the cache expired.
```

File: src/Metadata/CacheItems.php [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+{
+    /**
+     * @var array<string, CacheItem>
+     */
+    private array $items = [];
+
+    /**
+     * Checks whether the dependency has been changed and whether the cache expires.
```
Comment (samdark) [inline src/Metadata/CacheItems.php:22]:
```suggestion
     * Checks whether the dependency has been changed or whether the cache expired.
```

File: src/Metadata/CacheItems.php [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+     *
+     * @param string $key The key that identifies the cache item.
+     * @param float $beta The value for calculating the range that is used for "Probably early expiration" algorithm.
+     * @param CacheInterface $cache The actual cache handler.
+     *
+     * @return bool Whether the dependency has been changed and whether the cache expires.
```
Comment (samdark) [inline src/Metadata/CacheItems.php:28]:
```suggestion
     * @return bool Whether the dependency has been changed or whether the cache expired.
```

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**